include/swift/AST/Decl.h

```diff
@@ -3562,11 +3562,16 @@ class ClassDecl final : public NominalTypeDecl {
   void createObjCMethodLookup();
 
   struct {
+    /// The superclass decl and a bit to indicate whether the
+    /// superclass was computed yet or not.
+    llvm::PointerIntPair<ClassDecl *, 1, bool> SuperclassDecl;
+
     /// The superclass type and a bit to indicate whether the
     /// superclass was computed yet or not.
-    llvm::PointerIntPair<Type, 1, bool> Superclass;
+    llvm::PointerIntPair<Type, 1, bool> SuperclassType;
```

Comment:
Memory-usage-wise, do we need both? Seems like storing `PointerUnion<ClassDecl *, Type>` inside the PointerIntPair would be sufficient.

Reply:
The cached superclass `Decl` and superclass `Type ` are handled by different conceptual layers (name lookup vs. type resolution), so they're really different pieces of state. I don't think the benefits of folding them together are worth the risk of unpredictable behavior.

Reply:
Yeah. Even in a codebase with 10,000 classes and protocols, you're looking at an 80kb savings.